contracts/truefi2/CrvBaseRateOracle.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.6.10;
+
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import "../truefi/interface/ICurve.sol";
+
+// prettier-ignore
+contract CrvBaseRateOracle {
+    using SafeMath for uint256;
+    uint8 public constant BUFFER_SIZE = 7;
+
+    ICurve public curve;
+
+    struct HistoricalRatesBuffer {
+        uint256[BUFFER_SIZE] baseRates;
+        uint256[BUFFER_SIZE] timestamps;
+        uint8 insertIndex;
+    }
+
+    HistoricalRatesBuffer public histBuffer;
+
+    constructor(ICurve _curve) public {
+        curve = _curve;
+    }
+
+    function getHistBuffer() public view returns (uint256[BUFFER_SIZE] memory, uint256[BUFFER_SIZE] memory, uint8) {
+        return (histBuffer.baseRates, histBuffer.timestamps, histBuffer.insertIndex);
+    }
+
+    function updateRate() public {
+        uint8 iidx = histBuffer.insertIndex;
+        histBuffer.timestamps[iidx] = block.timestamp;
+        histBuffer.baseRates[iidx] = curve.get_virtual_price();
+        histBuffer.insertIndex = (iidx + 1) % BUFFER_SIZE;
+    }
+
+    function calculateAverageRate() public view returns (uint256) {
```

Comment:
Let's make `calculateAverageRate` accept `curCrvBaseRate` instead of latest measurement. t_n would then be block.timestamp and rates would be more dynamic